Crash when sending HNZ frame

**Describe the bug**
When trying to send a messages on the HNZ socket, if said socket was closed by the HNZ device, no error handling is performed and a SIGPIPE signal is emitted, resulting in the whole program shutting down.

**To Reproduce**
Steps to reproduce the behavior:
1. Build and run the unit tests of the fledge-south-hnz repository manually in a docker following the same steps as in the [ci configuration](https://github.com/fledge-power/fledge-south-hnz/blob/0d16a442ee5dcfa9c126e9afaf6589d5374fcb90/.github/workflows/ci.yml) up to `make hnz_coverage_sonar`
2. Notice the following logs:
```
#27 69.71 [----------] 4 tests from HNZTest
#27 69.71 [ RUN      ] HNZTest.TCPConnectionTwoPathOK
#27 86.72 Adresse de connexion : 0.0.0.0
#27 86.72 Adresse de connexion : 0.0.0.0
#27 86.72 [HNZ Server] Connection OK !!
#27 86.72 [HNZ Server] Connection OK !!
#27 86.72 [       OK ] HNZTest.TCPConnectionTwoPathOK (17005 ms)
#27 86.72 [ RUN      ] HNZTest.ReceivingMessagesTwoPath
#27 132.7 make[3]: *** [CMakeFiles/hnz_coverage_sonar.dir/build.make:58: CMakeFiles/hnz_coverage_sonar] Broken pipe
#27 132.7 make[2]: *** [CMakeFiles/Makefile2:152: CMakeFiles/hnz_coverage_sonar.dir/all] Error 2
#27 132.7 make[1]: *** [CMakeFiles/Makefile2:159: CMakeFiles/hnz_coverage_sonar.dir/rule] Error 2
#27 132.7 make: *** [Makefile:155: hnz_coverage_sonar] Error 2
```

**Expected behavior**
The error is handled in a clean way and does not result in a program crash.

**Desktop:**
 - OS: Windows 10
 - Browser: Firefox 112
 - Fledge: v2.1.0

**Additional context**
Attaching a debugger to the main test exacutable `fledge-south-hnz/build/tests/RunTests` produces the following logs and stack trace:
```c++
[----------] 4 tests from HNZTest
[ RUN      ] HNZTest.ReceivingMessagesTwoPath
[New Thread 0x7f03b7b88700 (LWP 17)]
[New Thread 0x7f03b7387700 (LWP 18)]
Fledge fledge[13]: INFO: Reading json config string...
Fledge fledge[13]: INFO: Json config parsed successsfully.
Fledge fledge[13]: WARNING: [Path A - active] Going to HNZ connection state... Waiting for a SARM.
Fledge fledge[13]: WARNING: [Path B - passive] Going to HNZ connection state... Waiting for a SARM.
Fledge fledge[13]: INFO: Starting HNZ south plugin...
[New Thread 0x7f03b6b86700 (LWP 19)]
Fledge fledge[13]: INFO: [Path A - active] Connecting to PA on 0.0.0.0 (6005)... [1/5]
Adresse de connexion : 0.0.0.0
[New Thread 0x7f03b5b84700 (LWP 20)]
[New Thread 0x7f03b6385700 (LWP 21)]
[Thread 0x7f03b7b88700 (LWP 17) exited]
Fledge fledge[13]: INFO: [Path A - active]Connected to 0.0.0.0 (6005).
[New Thread 0x7f03b5383700 (LWP 22)]
Fledge fledge[13]: DEBUG: [Path A - active] HNZ Connection Management thread running
Fledge fledge[13]: WARNING: [Path A - active] Listening for data...
Fledge fledge[13]: INFO: [Path A - active] SARM sent [1 / 5]
[New Thread 0x7f03b4b82700 (LWP 23)]
Fledge fledge[13]: DEBUG: HNZ Connection starting...
Fledge fledge[13]: INFO: [Path B - passive] Connecting to PA on 0.0.0.0 (6006)... [1/5]
Adresse de connexion : 0.0.0.0
[New Thread 0x7f03a7fff700 (LWP 24)]
[New Thread 0x7f03a77fe700 (LWP 25)]
[New Thread 0x7f03a6ffd700 (LWP 26)]
Fledge fledge[13]: INFO: [Path B - passive]Connected to 0.0.0.0 (6006).
[Thread 0x7f03b7387700 (LWP 18) exited]
[New Thread 0x7f03b7b88700 (LWP 27)]
[New Thread 0x7f03a67fc700 (LWP 28)]
Fledge fledge[13]: WARNING: [Path B - passive] Listening for data...
Fledge fledge[13]: DEBUG: [Path B - passive] HNZ Connection Management thread running
Fledge fledge[13]: DEBUG: [Path A - active] 5 15 8 137
Fledge fledge[13]: INFO: [Path A - active] Received SARM
Fledge fledge[13]: INFO: [Path B - passive] SARM sent [1 / 5]
Fledge fledge[13]: INFO: [Path A - active] UA sent
Fledge fledge[13]: DEBUG: [Path A - active] 7 99 210 19
Fledge fledge[13]: INFO: [Path A - active] Received UA
Fledge fledge[13]: DEBUG: [Path A - active] HNZ Connection initialized !!
Fledge fledge[13]: WARNING: [Path A - active] Time setting sent : 9/5/23
Fledge fledge[13]: WARNING: [Path A - active] Time setting sent : mod10m = 52 and 10ms frac = 43100 (8h40m and 431s 0ms
Fledge fledge[13]: WARNING: [Path A - active] GI (General Interrogation) request sent
Fledge fledge[13]: INFO: [Path A - active] BULLE sent
[HNZ Server] Connection OK !!
[Thread 0x7f03b7b88700 (LWP 27) exited]
Fledge fledge[13]: INFO: [Path B - passive] SARM sent [2 / 5]
[New Thread 0x7f03b7b88700 (LWP 29)]
Fledge fledge[13]: DEBUG: [Path A - active] 7 33 196 114
Fledge fledge[13]: WARNING: [Path A - active] RR received (f = 0, nr = 1)
[New Thread 0x7f03b7387700 (LWP 30)]
Fledge fledge[13]: DEBUG: [Path B - passive] 5 15 8 137
Fledge fledge[13]: INFO: [Path B - passive] Received SARM
Fledge fledge[13]: INFO: [Path B - passive] UA sent
Fledge fledge[13]: DEBUG: [Path A - active] 7 65 194 17
Fledge fledge[13]: WARNING: [Path A - active] RR received (f = 0, nr = 2)
Fledge fledge[13]: DEBUG: [Path A - active] 7 97 192 48
Fledge fledge[13]: WARNING: [Path A - active] RR received (f = 0, nr = 3)
Fledge fledge[13]: DEBUG: [Path B - passive] 7 99 210 19
Fledge fledge[13]: INFO: [Path B - passive] Received UA
Fledge fledge[13]: DEBUG: [Path B - passive] HNZ Connection initialized !!
Fledge fledge[13]: DEBUG: [Path A - active] 7 129 206 215
Fledge fledge[13]: WARNING: [Path A - active] RR received (f = 0, nr = 4)
Fledge fledge[13]: DEBUG: [Path B - passive] 7 99 210 19
Fledge fledge[13]: INFO: [Path B - passive] Received UA
Fledge fledge[13]: DEBUG: [Path B - passive] HNZ Connection initialized !!
Fledge fledge[13]: INFO: [Path B - passive] BULLE sent
[HNZ Server] Connection OK !!
[Thread 0x7f03b7387700 (LWP 30) exited]
[New Thread 0x7f03b7387700 (LWP 31)]
Fledge fledge[13]: DEBUG: [Path A - active] 5 0 11 51 40 54 242 22 161
Fledge fledge[13]: INFO: [Path A - active] Received an information frame (ns = 0, p = 0, nr = 0)
Fledge fledge[13]: INFO: [Path A - active] Received TSCE
Fledge fledge[13]: DEBUG: [Path B - passive] 7 33 196 114
Fledge fledge[13]: WARNING: [Path B - passive] RR received (f = 0, nr = 1)
[HNZ Server] TSCE sent on both path
Fledge fledge[13]: DEBUG: [Path A - active] [11 51 40 54 242]
Fledge fledge[13]: WARNING: [Path A - active] Ignoring the RR, NR (=0) is invalid. Current NRR : 5
Fledge fledge[13]: DEBUG: [Path B - passive] 5 0 11 51 40 54 242 22 161
Fledge fledge[13]: INFO: [Path B - passive] Received an information frame (ns = 0, p = 0, nr = 0)
Fledge fledge[13]: WARNING: [Path B - passive] Ignoring the RR, NR (=0) is invalid. Current NRR : 2
Fledge fledge[13]: INFO: [Path B - passive] RR sent
Fledge fledge[13]: INFO: [Path A - active] RR sent
Fledge fledge[13]: INFO: Pushing to Fledge a TSCE
Fledge fledge[13]: DEBUG: Send to fledge TSCE with station address = 1, message address = 511, value = 1, valid = 0ts = 14066, iv = 0, c = 0, s0
ingestCallback called -> asset: (TS1)
  number of readings: 1
Fledge fledge[13]: DEBUG: [Path A - active] 5 15 8 137
Fledge fledge[13]: INFO: [Path A - active] Received SARM
Fledge fledge[13]: WARNING: [Path A - active] Going to HNZ connection state... Waiting for a SARM.
Fledge fledge[13]: INFO: [Path A - active] UA sent
Fledge fledge[13]: INFO: [Path A - active] SARM sent [1 / 5]
Fledge fledge[13]: INFO: [Path A - active] SARM sent [2 / 5]
Fledge fledge[13]: INFO: [Path B - passive] BULLE sent
Fledge fledge[13]: DEBUG: [Path B - passive] 7 65 194 17
Fledge fledge[13]: WARNING: [Path B - passive] RR received (f = 0, nr = 2)
Fledge fledge[13]: INFO: [Path A - active] SARM sent [3 / 5]
Fledge fledge[13]: INFO: [Path A - active] SARM sent [4 / 5]
Fledge fledge[13]: INFO: [Path A - active] SARM sent [5 / 5]
Fledge fledge[13]: INFO: [Path B - passive] BULLE sent
Fledge fledge[13]: DEBUG: [Path B - passive] 7 97 192 48
Fledge fledge[13]: WARNING: [Path B - passive] RR received (f = 0, nr = 3)
Fledge fledge[13]: WARNING: [Path A - active] The maximum number of SARM was reached.
Fledge fledge[13]: ERROR: Switching active and passive path.
Fledge fledge[13]: ERROR: New active path is [Path B - active]
Fledge fledge[13]: INFO: [Path A - passive] SARM sent [1 / 5]
Fledge fledge[13]: INFO: [Path A - passive] SARM sent [2 / 5]
Fledge fledge[13]: INFO: [Path A - passive] SARM sent [3 / 5]
Fledge fledge[13]: INFO: [Path B - active] BULLE sent
Fledge fledge[13]: DEBUG: [Path B - active] 7 129 206 215
Fledge fledge[13]: WARNING: [Path B - active] RR received (f = 0, nr = 4)
Fledge fledge[13]: INFO: [Path A - passive] SARM sent [4 / 5]
[HNZ Server] TSCE sent on path B
Fledge fledge[13]: DEBUG: [Path B - active] 5 2 11 51 40 54 242 64 169
Fledge fledge[13]: INFO: [Path B - active] Received an information frame (ns = 1, p = 0, nr = 0)
Fledge fledge[13]: INFO: [Path B - active] Received TSCE
Fledge fledge[13]: DEBUG: [Path B - active] [11 51 40 54 242]
Fledge fledge[13]: WARNING: [Path B - active] Ignoring the RR, NR (=0) is invalid. Current NRR : 5
Fledge fledge[13]: INFO: [Path B - active] RR sent
Fledge fledge[13]: INFO: [Path A - passive] SARM sent [5 / 5]
Fledge fledge[13]: INFO: Pushing to Fledge a TSCE
Fledge fledge[13]: DEBUG: Send to fledge TSCE with station address = 1, message address = 511, value = 1, valid = 0ts = 14066, iv = 0, c = 0, s0
ingestCallback called -> asset: (TS1)
  number of readings: 1
[Thread 0x7f03b6385700 (LWP 21) exited]
[Thread 0x7f03b7b88700 (LWP 29) exited]
[Thread 0x7f03b7387700 (LWP 31) exited]
[Thread 0x7f03a6ffd700 (LWP 26) exited]
Fledge fledge[13]: INFO: Starting shutdown of HNZ plugin
Fledge fledge[13]: DEBUG: HNZ Connection stopping...
Fledge fledge[13]: WARNING: [Path A - passive] The maximum number of SARM was reached.
[Thread 0x7f03b4b82700 (LWP 23) exited]
[Thread 0x7f03b6b86700 (LWP 19) exited]
Fledge fledge[13]: DEBUG: [Path B - active] HNZ Connection stopping...
Fledge fledge[13]: INFO: [Path A - passive] SARM sent [1 / 5]
Fledge fledge[13]: DEBUG: [Path B - active] Waiting for the connection thread
[Thread 0x7f03a77fe700 (LWP 25) exited]
[Thread 0x7f03a7fff700 (LWP 24) exited]
--Type <RET> for more, q to quit, c to continue without paging--

Thread 7 "RunTests" received signal SIGPIPE, Broken pipe.
[Switching to Thread 0x7f03b5383700 (LWP 22)]
0x00007f03b8d677dc in send () from /lib/x86_64-linux-gnu/libpthread.so.0
(gdb) bt
#0  0x00007f03b8d677dc in send () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f03b869637b in TcpConnexion::vSend(unsigned char*, int) () from /tmp/libhnz/src/hnz/liblibhnz.so
#2  0x00007f03b8696ac3 in VoieHNZ::vEnvoiTrameVersHnz() () from /tmp/libhnz/src/hnz/liblibhnz.so
#3  0x00007f03b8696911 in VoieHNZ::vEnvoiTrameVersHnz(MSG_TRAME*) () from /tmp/libhnz/src/hnz/liblibhnz.so
#4  0x00007f03b869ddd0 in HNZClient::sendFr(MSG_TRAME*) () from /tmp/libhnz/src/hnz/liblibhnz.so
#5  0x00007f03b869dd80 in HNZClient::createAndSendFr(unsigned char, unsigned char*, int) () from /tmp/libhnz/src/hnz/liblibhnz.so
#6  0x000056239b5a0b88 in HNZPath::m_sendSARM (this=0x56239bb63870) at /tmp/fledge-south-hnz/src/hnzpath.cpp:439
#7  0x000056239b5a6d41 in HNZPath::m_manageHNZProtocolConnection (this=0x56239bb63870) at /tmp/fledge-south-hnz/src/hnzpath.cpp:155
#8  0x000056239b5a9b75 in std::__invoke_impl<void, void (HNZPath::*)(), HNZPath*> (__t=<optimized out>, __f=<optimized out>) at /usr/include/c++/9/bits/invoke.h:89
#9  std::__invoke<void (HNZPath::*)(), HNZPath*> (__fn=<optimized out>) at /usr/include/c++/9/bits/invoke.h:95
#10 std::thread::_Invoker<std::tuple<void (HNZPath::*)(), HNZPath*> >::_M_invoke<0ul, 1ul> (this=<optimized out>) at /usr/include/c++/9/thread:244
#11 std::thread::_Invoker<std::tuple<void (HNZPath::*)(), HNZPath*> >::operator() (this=<optimized out>) at /usr/include/c++/9/thread:251
#12 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (HNZPath::*)(), HNZPath*> > >::_M_run (this=<optimized out>) at /usr/include/c++/9/thread:195
#13 0x00007f03b8572de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#14 0x00007f03b8d5c609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#15 0x00007f03b83ac133 in clone () from /lib/x86_64-linux-gnu/libc.so.6
(gdb)   
```

--- a/src/hnz/TcpConnexion.cpp
+++ b/src/hnz/TcpConnexion.cpp
@@ -83,7 +83,13 @@ void TcpConnexion::iTCPConnecteServeur(int port) {
 
 void TcpConnexion::vSend(unsigned char *buffer, int buffSize) {
   // vCSEcriture(true) ;
-  send(socketfd, buffer, buffSize, NULL);
+  // Do not send SIGPIPE in case of broken pipe
+  ssize_t sent = send(socketfd, buffer, buffSize, MSG_NOSIGNAL);
+  if(sent == -1){
+    // Display error if any
+    perror("send:");
+    perror(strerror(errno));
+  }
   // vCSEcriture(false) ;
 }
 

